gh-117114: Add `isdevdrive` to `posixpath`

nineteendo opened the pull request:
<!-- gh-issue-number: gh-117114 -->
* Issue: gh-117114
<!-- /gh-issue-number -->


<!-- readthedocs-preview cpython-previews start -->
----
📚 Documentation preview 📚: https://cpython-previews--117115.org.readthedocs.build/

<!-- readthedocs-preview cpython-previews end -->

nineteendo committed:
Same interface for `posixpath` & `ntpath`

blurb-it[bot] committed:
📜🤖 Added by blurb_it.

nineteendo committed:
Decouple logic for `splitdrive` from `splitroot`

nineteendo committed:
Fixed `splitdrive` for "/:/foo"

zware reviewed (dismissed):
> I moved the test from `test_splitroot` to `test_splitdrive`:

You're going to need a really good justification for that: the test isn't just moved, it's changed.  See [the original issue](https://github.com/python/cpython/pull/101363#issuecomment-1407403056) for more context on that test (cc @barneygale and @eryksun).

I have not actually reviewed the rest of this PR, but this strikes me as a blocking issue.

barneygale commented on Lib/ntpath.py:
```diff
@@ -179,45 +206,19 @@ def splitroot(p):
     """
-    p = os.fspath(p)
+    drive, p = splitdrive(p)
     if isinstance(p, bytes):
```
Repeats work already done in `splitdrive()`

eryksun replied:
@barneygale, is there a relevant discussion for this reversion to doing all the work in `splitdrive()`? I thought the plan was to implement a fast version of `splitroot()` in C, which would be used by `splitdrive()` and other functions.

nineteendo replied:
@eryksun, no. I wasn't aware of such plans. But the way it's written right now, the logic of splitting the drive and root is intertwined. Logically you need to split the drive, and the root too if the remaining path starts with a (back)slash.

eryksun replied:
The implementations of `os.path` functions call `splitroot()`, not `splitdrive()`. This change will make all of that code slower. The extra cost to split out the root is minimal in context, but Python function calls are always expensive.

barneygale commented on Lib/ntpath.py:
```diff
@@ -179,45 +206,19 @@ def splitroot(p):
-        unc_prefix = '\\\\?\\UNC\\'
-        empty = ''
     normp = p.replace(altsep, sep)
     if normp[:1] == sep:
```
Repeats work already done in `splitdrive()`

barneygale commented on Lib/posixpath.py:
```diff
@@ -168,23 +168,14 @@ def splitroot(p):
-    sep = _get_sep(p)
-    i = p.rfind(sep) + 1
-    return p[i:]
+    return split(p)[1]
```
Slower because it does more work.

barneygale commented on Lib/posixpath.py:
```diff
@@ -168,23 +168,14 @@ def splitroot(p):
-    if head and head != sep*len(head):
-        head = head.rstrip(sep)
-    return head
+    return split(p)[0]
```
Slower because it does more work.

barneygale commented on Lib/posixpath.py:
```diff
@@ -196,16 +187,16 @@ def isjunction(path):
+def isreserved(path):
+    """Return true if the pathname is reserved by the system.
+    Always returns False on posix"""
+    return False
```
This was deliberately excluded from the `posixpath` interface - see https://github.com/python/cpython/pull/95486#issuecomment-1341553054

nineteendo committed:
Make requested changes

nineteendo committed:
Revert `splitroot`

nineteendo committed:
Remove comment

Eclips4 commented on Misc/NEWS.d/next/Core and Builtins/2024-03-21-09-57-57.gh-issue-117114.Qu-p55.rst:
```diff
@@ -0,0 +1 @@
+Added posix implementation for isdevdrive.
```
```suggestion
Added POSIX implementation for :func:`os.path.isdevdrive`.
```

nineteendo committed:
Update Misc/NEWS.d/next/Core and Builtins/2024-03-21-09-57-57.gh-issue-117114.Qu-p55.rst

Co-authored-by: Kirill Podoprigora <kirill.bast9@mail.ru>

nineteendo committed:
Allow checking support for Dev Drives / junctions

You can now check if a platform supports Dev Drives / junctions:

- `ntpath.isdevdrive` is not `genericpath.isdevdrive`, `posixpath.isdevdrive` is not `genericpath.isdevdrive`
- `ntpath.isjunction` is not `genericpath.isjunction`, `posixpath.isjunction` is not `genericpath.isjunction`

zooba commented on Doc/library/os.path.rst:
```diff
@@ -324,10 +324,13 @@ the :mod:`glob` module.)
    for information on enabling and creating Dev Drives.
 
-   .. availability:: Windows.
+   .. availability:: Unix, Windows.
```
Just remove this line.

zooba commented on Misc/NEWS.d/next/Core and Builtins/2024-03-21-09-57-57.gh-issue-117114.Qu-p55.rst:
```diff
@@ -0,0 +1 @@
+Added POSIX implementation for :func:`os.path.isdevdrive`.
```
```suggestion
Make :func:`os.path.isdevdrive` available on all platforms. For those that do not offer Dev Drives, it will always return ``False``.
```

zooba reviewed (commented):
Moving into genericpath is good, though I wouldn't suggest the `is` test to see whether Dev Drives are "real" or not (e.g. to suggest that something may run quicker with one set up). May as well just do `os.name == "nt"` at that point and mention updating to Windows 11, let the user figure it out.

nineteendo committed:
Update Misc/NEWS.d/next/Core and Builtins/2024-03-21-09-57-57.gh-issue-117114.Qu-p55.rst

Co-authored-by: Steve Dower <steve.dower@microsoft.com>

nineteendo committed:
added link to `genericpath.py`
